api/billingd/service/service.go

```diff
@@ -0,0 +1,708 @@
+				"instance_writes.units":     0,
+				"instance_writes.sub_units": 0,
+			}}); err != nil {
+			return nil, err
+		}
+	}
+
+	log.Debugf("updated subscription with status '%s' for %s", req.Status, req.CustomerId)
+	return &pb.UpdateCustomerSubscriptionResponse{}, nil
+}
+
+func (s *Service) RecreateCustomerSubscription(ctx context.Context, req *pb.RecreateCustomerSubscriptionRequest) (
+	*pb.RecreateCustomerSubscriptionResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	r := s.db.FindOne(ctx, bson.M{"_id": req.CustomerId})
+	if r.Err() != nil {
+		return nil, r.Err()
+	}
+	var doc Customer
+	if err := r.Decode(&doc); err != nil {
+		return nil, err
+	}
+	if err := common.StatusCheck(doc.Status); err == nil {
+		return nil, common.ErrSubscriptionExists
+	} else if !errors.Is(err, common.ErrSubscriptionCanceled) {
+		return nil, err
+	}
+	if err := s.createSubscription(&doc); err != nil {
+		return nil, err
+	}
+	if _, err := s.db.UpdateOne(ctx, bson.M{"_id": req.CustomerId}, bson.M{
+		"$set": bson.M{
+			"status":          doc.Status,
+			"period":          doc.Period,
+			"stored_data":     doc.StoredData,
+			"network_egress":  doc.NetworkEgress,
+			"instance_reads":  doc.InstanceReads,
+			"instance_writes": doc.InstanceWrites,
+		}}); err != nil {
+		return nil, err
+	}
+
+	log.Debugf("recreated subscription for %s", req.CustomerId)
+	return &pb.RecreateCustomerSubscriptionResponse{}, nil
+}
+
+func (s *Service) DeleteCustomer(ctx context.Context, req *pb.DeleteCustomerRequest) (
+	*pb.DeleteCustomerResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	if _, err := s.stripe.Customers.Del(req.CustomerId, nil); err != nil {
+		return nil, err
+	}
+	if _, err := s.db.DeleteOne(ctx, bson.M{"_id": req.CustomerId}); err != nil {
+		return nil, err
+	}
+	log.Debugf("deleted customer %s", req.CustomerId)
+	return &pb.DeleteCustomerResponse{}, nil
+}
+
+func (s *Service) IncStoredData(ctx context.Context, req *pb.IncStoredDataRequest) (*pb.IncStoredDataResponse, error) {
+	lck := s.semaphores.Get(customerLock(req.CustomerId))
+	lck.Acquire()
+	defer lck.Release()
+
+	r := s.db.FindOne(ctx, bson.M{"_id": req.CustomerId})
+	if r.Err() != nil {
+		return nil, r.Err()
+	}
+	var doc Customer
+	if err := r.Decode(&doc); err != nil {
+		return nil, err
+	}
+	if err := common.StatusCheck(doc.Status); err != nil {
+		return nil, err
+	}
+	usage, err := s.handleUsage(
+		ctx,
+		doc.ID,
+		doc.StoredData,
+		"stored_data",
+		req.IncSize,
+		StoredDataUnitSize,
+		StoredDataFreeUnits,
+		doc.Billable,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugf(
+		"%s period data: units=%d total=%d free=%d",
+		req.CustomerId,
+		usage.Units,
+		usage.Total,
+		usage.Free,
+	)
+	return &pb.IncStoredDataResponse{
+		Period:     periodToPb(doc.Period),
+		StoredData: usage,
+	}, nil
+}
+
+func (s *Service) handleUsage(
```

Comment:
This is the logic that handles incrementing a usage value and when to fire off an request to stripe.